### ISSUE TITLE
Tracker music support

### ISSUE THREAD

**Issue by est31** [minetest/minetest#3878]
@tenplus1 and others on IRC have brought up the need for music on minetest servers. One of its downsides is its big size. This can be fought with by using tracker music, as they pointed out. The most popular format there is [MOD](https://en.wikipedia.org/wiki/MOD_%28file_format%29).

It essentially consists of two parts: one that defines samples, and a second one that defines how and when they are played. This way it can encode music very well scaling (large constant + x \* small factor).

This would be a big leap because right now the sole supported file format is OGG.

Related to #1409 (@kaeza said similar things over at that issue).

**Comment by est31** [comment]
Some libraries:

* http://mikmod.sourceforge.net/
* http://lib.openmpt.org/libopenmpt (has no debian package, we would have to nudge the minetest packager)
* http://modplug-xmms.sourceforge.net/ (has debian package. A libopenmt developer [criticised this library](https://trac.videolan.org/vlc/ticket/13055))

**Comment by Fixer-007** [comment]
Yes, this can be handy, there are a lot of 8bit masterpieces.

**Comment by Cat5TV** [comment]
We'd love to see this added with support for the various formats, in particular MOD, S3M and XM. I will gladly contribute to the music library and assist with ideas, as I've worked with mod trackers since I was but a wee lad.

**Comment by paramat** [comment]
MOD is a good idea, and is also simple and a little retro so seems suitable for MT.

**Comment by everamzah** [comment]
"Mod" is a general term used to refer to modules such as .mod, .s3m, .xm, and .it.

**Comment by 4aiman** [comment]
This would be the third time the subject being brought up. :+1: 

But right now the *.ogg extension is hardcoded within the engine.
When one asks to load a ```soundname``` the engine searches for multiple ogg files ```soundnameX.ogg```.
So some changes to how the engine loads audio data should be applied in the first place.

**Comment by Wuzzy2** [comment]
I like the general idea, but I don't think MOD (“.mod”) is the best choice here. It is a bit old and not very powerful, maybe we should go with some of the newer ones, like XM and IT. Or we just support all of them. xD

**Comment by everamzah** [comment]
One library will support .mod, .s3m, .xm, and .it.

**Comment by paramat** [comment]
Yeah, i'm only using the term MOD because i don't know what i'm talking about :)

**Comment by lisacvuk** [comment]
+1!

**Comment by lightonflux** [comment]
> But right now the *.ogg extension is hardcoded within the engine.
>
> When one asks to load a soundname the engine searches for multiple ogg files soundnameX.ogg.

Can't the engine assume .ogg if no extension is provided? For compatibility. And for all the new formats you have to specify the extensions?

**Comment by Zeno-** [comment]
(sound)-(pro-)(etc-)tracker music is great

**Comment by benrob0329** [comment]
This seems to play my small collection of MODs quite nicely. 

http://xmp.sourceforge.net/

**Comment by benrob0329** [comment]
Forgot to mention, it has both a Debian and a Fedora package.

**Comment by Calinou** [comment]
> This seems to play my small collection of MODs quite nicely.
>
> http://xmp.sourceforge.net/

It has LGPLv2.1+ license, so we should be able to use it.

**Comment by everamzah** [comment]
https://github.com/minetest/minetest/issues/3878

@Fixer-007:
>mod, xm, it, s3m is much better for my ears

Quality of rendering will depend on the library used.  Some certainly do sound better than others!

**Comment by HybridDog** [comment]
There are some, maybe not up to date, tests:
https://github.com/schismtracker/schismtracker/wiki/Player-abuse-tests

**Comment by benrob0329** [comment]
That's only for Impulse Tracker, whatever library is chosen should be a good middle ground between the various formats.

**Comment by everamzah** [comment]
Listening to a mod in its editor - .XMs played inside FastTracker II, for example - will always sound better than using a support library.

**Comment by benrob0329** [comment]
I really like this 8 bit theme.

![screenshot_2016-07-19_20-16-28](https://cloud.githubusercontent.com/assets/19180682/16971062/6ef62d44-4dee-11e6-83ea-4460b0c353e5.png)

**Comment by kaeza** [comment]
>Listening to a mod in its editor - .XMs played inside FastTracker II, for example - will always sound better than using a support library.

That ultimately depends on the playback implementation. As always, there's a tradeoff between performance and quality of sound.

XM/MOD/whatever is still much better than MIDI, IMHO, for reasons I've outlined in [#1409](https://github.com/minetest/minetest/issues/1409#issuecomment-46880187) (except maybe for the actual music file size).

**Comment by everamzah** [comment]
```
debian@stable:~$ apt search xm s3m it mod
Sorting... Done
Full Text Search... Done
<snip>
libdumb1/stable 1:0.9.3-6+b1 amd64
  dynamic universal music bibliotheque

libmikmod3/stable,now 3.3.7-1 amd64 [installed,automatic]
  Portable sound library

libxmp4/stable 4.3.0-1 amd64
  module file rendering library
```

**Comment by Cat5TV** [comment]
By closing this does it mean there's no hope for it? I was really, really hoping for this feature as music is sorely lacking in Minetest and using OGG will simply slow the Media download way too much due to the way Minetest works with media.

**Comment by benrob0329** [comment]
:-(

**Comment by red-001** [comment]
@Cat5TV http://irc.minetest.net/minetest-dev/2017-04-25#i_4896485

**Comment by Fixer-007** [comment]
I disagree about closing this and other PRs, closed issues/prs by est31 should be reviewed and reopened again.

**Comment by SmallJoker** [comment]
Too popular to be closed.

**Comment by Cat5TV** [comment]
Thank you @SmallJoker - IMHO, this one is too _important_ to be closed (until implemented).

**Comment by ryvnf** [comment]
I support this idea. I think a format that doesn't require the music file to also include the samples/soundfont would be best so lots of different music can be sent from server to client with minimal bandwidth usage.

I personally like MIDI because it only encodes the sheet music itself, and that you can get slightly different timbre/quality by changing the soundfont or the engine. Maybe it would also be possible to implement MIDI with an API to also specify custom soundfonts, so people writing mods still have a lot of control to get the specific sound they are looking.

One thing that would be cool would be if it is possible to create a mod for players to create music on multiplayer servers and share it with each other (like Minecraft music discs, but recorded by players). Sort of like books but for music.

**Comment by tacotexmex** [comment]
I agree with the argument for MIDI support. Bundling the Minetest client with a good default soundfont would enable near infinite musical variation and experimentation without increasing bandwith. 

The default soundfont could be implemented like another resource (such as textures and sounds) so makers would be able to overwrite or extend it.

**Comment by hecktest** [comment]
I'd like to mention that on Windows, having an option to use installed MIDI devices is a must. The user can then redirect the MIDI to a superior renderer like BASSMIDI, or even to an actual physical device that they own.

In other words, the support must be no worse than what ZDoom or DOSbox provide.

**Comment by benrob0329** [comment]
> I'd like to mention that on Windows, having an option to use installed MIDI devices is a must. The user can then redirect the MIDI to a superior renderer like BASSMIDI, or even to an actual physical device that they own.
> 
> In other words, the support must be no worse than what ZDoom or DOSbox provide.

If Minetest only allows one Soundfont to be used, then I support this (although having to pass MIDI data to either Windows' MIDI routing, or ALSA, or whatever Mac uses might be a bit too much platform dependent code). If multiple Soundfonts are allowed however, then this could break everything rather easily by blanket-redirecting all MIDI data despite different contexts being used.

**Comment by IvanTurgenev** [comment]
it is possible to stream music as of right now?

**Comment by Calinou** [comment]
@IvanTurgenev As far as I know, streaming audio isn't possible yet.

**Comment by rollerozxa** [comment]
Adding libopenmpt support to Minetest would be a very useful way of allowing game music that isn't absolutely massive in filesize. In particular the MO3 format would be perfect (Impulse Tracker format with compressed samples) - Bejeweled 2 used this format for its music and managed to fit 40 minutes of music with high quality samples in a little under 450kb.

**Comment by thevbw** [comment]
MIDI support would vastly expand the list of compatible music, but the licensing could be problematic if people start copying in random files from the internet

**Comment by NathanSalapat** [comment]
Licensing isn't the problem of the engine. People can already use media that they don't have legal rights to.

**Comment by thevbw** [comment]
> People can already use media that they don't have legal rights to.

Yes, but they're less likely to think about copyright if it's a MIDI.

> Licensing isn't the problem of the engine.

I'm not saying it is; I'm saying it could cause problems that may be associated with the engine. Might be worth adding a reminder somewhere.

**Comment by NathanSalapat** [comment]
> > People can already use media that they don't have legal rights to.
> 
> Yes, but they're less likely to think about copyright if it's a MIDI.
> 

Do you have any source for this claim? I see no reason why anybody should be any more inclined to steal a MIDI file than any other file.

**Comment by thevbw** [comment]
> source?

You can tell by the countless question threads online from people asking whether they can dodge copyright by using downloaded MIDI clones of songs.

**Comment by jeremyshannon** [comment]
I agree with Rolleroxza, Libopenmpt would be a great choice, it's an excellent library that supports just about everything, and is still pretty tiny.
